File: Backend/TrackrAPI/TrackrAPI/Services/Catalogo/EspecialidadService.cs
using TrackrAPI.Dtos.Catalogo;
using TrackrAPI.Models;
using TrackrAPI.Repositorys.Catalogo;
using System.Collections.Generic;

namespace TrackrAPI.Services.Catalogo
{
    public class EspecialidadService
    {
        private readonly IEspecialidadRepository especialidadRepository;
        private readonly EspecialidadValidatorService especialidadValidatorService;

        public EspecialidadService(
            IEspecialidadRepository especialidadRepository,
            EspecialidadValidatorService especialidadValidatorService)
        {
            this.especialidadRepository = especialidadRepository;
            this.especialidadValidatorService = especialidadValidatorService;
        }

        public EspecialidadFormularioConsultaDto? ConsultarParaFormulario(int idEspecialidad)
        {
            var especialidad = especialidadRepository.Consultar(idEspecialidad);

            if (especialidad is null)
            {
                return null;
            }

            var especialidadDto = new EspecialidadFormularioConsultaDto
            {
                IdEspecialidad = especialidad.IdEspecialidad,
                Nombre = especialidad.Nombre
            };

            return especialidadDto;
        }

        public IEnumerable<EspecialidadGridDto> ConsultarParaGrid()
        {
            var especialidades = especialidadRepository.Consultar();

            var especialidadesDto = especialidades
                .OrderBy(es => es.Nombre)
                .Select(es => new EspecialidadGridDto
                {
                    IdEspecialidad = es.IdEspecialidad,
                    Nombre = es.Nombre
                });

            return especialidadesDto;
        }

         public void Agregar(EspecialidadFormularioCapturaDto especialidadDto)
        {
            especialidadValidatorService.ValidarAgregar(especialidadDto);

            var especialidad = new Especialidad
            {
               /*  IdEspecialidad = especialidadDto.IdEspecialidad, */
                Nombre = especialidadDto.Nombre
            };

            especialidadRepository.Agregar(especialidad);
        }

        public void Editar(EspecialidadFormularioCapturaDto especialidadDto)
        {
            especialidadValidatorService.ValidarEditar(especialidadDto);

            var especialidad = especialidadRepository.ConsultarPorNombre(especialidadDto.Nombre)!;

            especialidad.Nombre = especialidadDto.Nombre;
            especialidadRepository.Editar(especialidad);
        }

        public void Eliminar(int idEspecialidad)
        {

            var especialidad = especialidadRepository.Consultar(idEspecialidad)!;

            especialidadRepository.Eliminar(especialidad!);
        }
    }
}

File: Frontend/trackr-administrador/src/app/shared/http/catalogo/especialidad.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Especialidad } from '@models/catalogo/especialidad';
import { EspecialidadFormularioConsultaDto } from '@dtos/catalogo/especialidad-formulario-consulta-dto';
import { EspecialidadGridDto } from '@dtos/catalogo/especialidad-grid-dto';
import { EspecialidadFormularioCapturaDto } from '@dtos/catalogo/especialidad-formulario-captura-dto';

@Injectable({
  providedIn: 'root',
})
export class EspecialidadService {
  private dataUrl = 'especialidad/';

  constructor(private http: HttpClient) {}

  consultarParaFormulario(idEspecialidad: number): Observable<EspecialidadFormularioConsultaDto> {
    return this.http.get<EspecialidadFormularioConsultaDto>(this.dataUrl + `formulario/${idEspecialidad}`);
  }

  consultarParaGrid(): Observable<EspecialidadGridDto[]> {
    return this.http.get<EspecialidadGridDto[]>(this.dataUrl + 'grid');
  }

  agregar(especialidad: EspecialidadFormularioCapturaDto): Observable<void>
  {
    return this.http.post<void>(this.dataUrl, especialidad);
  } 

  /* agregar(especialidad: EspecialidadFormularioCapturaDto): Observable<number>
  {
    return this.http.post<number>(this.dataUrl, 35);
  } */
  
  editar(especialidad: EspecialidadFormularioCapturaDto): Observable<void> {
    return this.http.put<void>(this.dataUrl, especialidad);
  }

  eliminar(idEspecialidad: number): Observable<void> {
    return this.http.delete<void>(this.dataUrl + `${idEspecialidad}`);
  }
}